Settings - Writing: add a toggle to Carousel so users can hide comment area

eliorivero opened the pull request:
Fixes #15224.

#### Changes proposed in this Pull Request:

* this PR introduces a new toggle in Jetpack's Writing settings, under the Carousel section, to hide the comment button, comment text box, and the list of comments.
* introduces a similar settings in Settings > Media
* refactors the Writing component from its class-based approach to use React hooks

#### Testing instructions:

- Go to WP Admin > Jetpack > Settings. In the Writing tab, locate the Carousel block
- Make sure there's a toggle to hide the comments area
- Click it to disable it
- Go to a post with a gallery where images are linked to the carousel
- Click on an image to launch it in the carousel
- Make sure the Comment button, the text box to add a comment, and the comment list are all hidden

### Before

<img width="678" alt="Screen Shot 2020-04-13 at 20 58 45" src="https://user-images.githubusercontent.com/1041600/79172043-fe813300-7dc9-11ea-8986-7eec82bf148f.png">

<img width="622" alt="Screen Shot 2020-04-13 at 20 42 46" src="https://user-images.githubusercontent.com/1041600/79171916-a2b6aa00-7dc9-11ea-94e6-a95e5c29a488.png">

### After

<img width="656" alt="Screen Shot 2020-04-13 at 20 41 42" src="https://user-images.githubusercontent.com/1041600/79171919-a34f4080-7dc9-11ea-80e7-3411c63a71c1.png">

<img width="617" alt="Screen Shot 2020-04-13 at 20 42 32" src="https://user-images.githubusercontent.com/1041600/79171917-a34f4080-7dc9-11ea-8c26-2776fd555b85.png">

### Comments on

<img width="1074" alt="Screen Shot 2020-04-13 at 21 06 05" src="https://user-images.githubusercontent.com/1041600/79172330-c75f5180-7dca-11ea-81d8-62529dcf8907.png">

### Comments off

<img width="1023" alt="Screen Shot 2020-04-13 at 20 51 10" src="https://user-images.githubusercontent.com/1041600/79171904-9d595f80-7dc9-11ea-89df-a7f5feec5454.png">

#### Proposed changelog entry for your changes:

* Carousel: new setting allows to hide the comments section in the image lightbox

eliorivero committed:
Carousel: allow to hide comment text box and list of comments.

Includes toggle for Jetpack React settings and checkbox for Settings > Media.

eliorivero committed:
Carousel: rewrite component using hooks. This results in less compiled code.

eliorivero committed:
Components - WritingMedia: refactor to consolidate some code so it passes CodeClimate verification.

eliorivero committed:
Add instruction to test new option in Carousel

eliorivero commented:
Addressed conflict after the introduction of https://github.com/Automattic/jetpack/commit/3010b47431d784ba4771188ca1da2637bb922d97#diff-f89b5798804790c524f02a00b0bc32a0

jeherve reviewed (changes requested):
You'll need to rebase this Pull Request as well. Like the other one, tests are failing because of WP Core changes and we have now updated `master` to take this into account.

Back to the PR itself, I am having trouble getting the toggle to behave properly in the Jetpack dashboard. When I first access the dashboard with a deactivated Carousel (the default behaviour on a brand new site), all the toggles are off. If I activate Carousel, the toggle you've added does not automatically appear as on. If I refresh the page, however, the toggle is on at that point.

eliorivero commented:
Got it. Will get back to this asap. Thanks @jeherve

eliorivero commented:
I've solved the conflicts.

The bug you mentioned seems to occur even with the master branch so it's not related to this PR but I'll try to fix it since it's not a good UX.

![carousel](https://user-images.githubusercontent.com/1041600/81997618-ad619a80-9626-11ea-90fa-13e26660d763.gif)

eliorivero committed:
Components - WritingMedia: fix Carousel  subtoggles not updating when Carousel was turned on.

Toggles to display EXIF data and comments didn't update according to its
value when the parent Carousel control was toggled on.

eliorivero commented:
Fixed in https://github.com/Automattic/jetpack/pull/15436/commits/e917f209967bbbc0569fc348b5de6f14d19b1a91

![carousel](https://user-images.githubusercontent.com/1041600/82001037-84de9e00-9630-11ea-8461-3ee0a89fe1a1.gif)

jeherve reviewed (approved):
This works well in my tests, and will make a great new feature for the Carousel. Thank you!

jeherve commented:
r207676-wpcom
